### ISSUE TITLE
Python client streaming race condition

### ISSUE THREAD

**Issue by kpayson64** [grpc/grpc#6360]
After creating stress test worker in #6359, the following test fairly reliably produces an error:
 
```PYTHONPATH=/workspace/grpc/src/python/grpcio:/workspace/grpc/src/python/gens/ python src/python/grpcio/tests/stress/client.py  --num_stubs_per_channel 10 --test_duration_secs 10 --test_cases client_streaming:100```

```Traceback (most recent call last):
  File "/workspace/grpc/src/python/grpcio/tests/stress/test_runner.py", line 66, in run
    test_case.test_interoperability(self._stub, None)
  File "/workspace/grpc/src/python/grpcio/tests/interop/methods.py", line 330, in test_interoperability
    _client_streaming(stub)
  File "/workspace/grpc/src/python/grpcio/tests/interop/methods.py", line 129, in _client_streaming
    response = stub.StreamingInputCall(requests, _TIMEOUT)
  File "/workspace/grpc/src/python/grpcio/grpc/framework/crust/implementations.py", line 124, in __call__
    protocol_options, metadata, request_iterator, self._pool)
  File "/workspace/grpc/src/python/grpcio/grpc/framework/crust/_calls.py", line 146, in blocking_stream_unary
    return next(rendezvous)
  File "/workspace/grpc/src/python/grpcio/grpc/framework/crust/_control.py", line 423, in next
    raise StopIteration()
StopIteration
Traceback (most recent call last):
  File "src/python/grpcio/tests/stress/client.py", line 132, in <module>
    run_test(_args())
  File "src/python/grpcio/tests/stress/client.py", line 120, in run_test
    raise exception_queue.get(block=True, timeout=args.test_duration_secs)
Exception: ('An exception occured during test TestCase.CLIENT_STREAMING', StopIteration())```

**Comment by nathanielmanistaatgoogle** [comment]
@kpayson64: How easily do you think you could turn this into a failing unit test?

**Comment by kpayson64** [comment]
I didn't do much investigation on this issue so it would be difficult to reproduce deterministically with without knowing the underlying root cause.

The stress test fails less often the shorter I run it for, so any naive multi-channel-per-stub test will end up being a flake test.

**Comment by croachrose** [comment]
I have been struggling with the error for the past couple of days. I cannot reliably reproduce but I have spent some time trying to figure out the cause. Here is what I know (correct me if I am wrong):

1. StopIteration should be thrown during a streaming RPC to stop iteration (generators or for loops)
2. In my case, I'm only using unary calls so StopIteration should never be thrown.
3. Somehow, the payload is None but _up_completion.kind is _Awaited.Kind.ARRIVED so the RPC has finished by the payload is missing?

I am still trying to gather more information. If it helps, I'm using the python module to issue RPCs to a golang service. I will see if I can verify that all the golang responses have a valid payload.

**Comment by nathanielmanistaatgoogle** [comment]
Thank you for the investigation @croachrose. I'm working on performance improvements that are likely to bypass wherever this defect is (which is why it's marked "to close"), but if this could be figured out it might be nice to fix the 0.14 development line.

**Comment by kpayson64** [comment]
Verified that this issue has been resolved with the latest GA Changes.  The fix will be included in the next release.
